File: Cpp/0424_longest_repeating_character_replacement.cpp
// https://leetcode.cn/problems/longest-repeating-character-replacement/

class Solution {
public:
    int characterReplacement(string s, int k) {
        int res = 0;
        int left = 0, right = 0;
        vector<int> cnt(26, 0);
        while (right < s.size()) {
            cnt[s[right] - 'A'] += 1;
            while (!isValid(cnt, k)) {
                cnt[s[left] - 'A'] -= 1;
                left += 1;
            }

            int tmp = right - left + 1;
            if (tmp > res) {
                res = tmp;
            }

            right += 1;
        }
        return res;
    }

    bool isValid(vector<int>& cnt, int k) {
        int max_cnt = 0, sum_cnt = 0;
        for (int i: cnt) {
            if (i > max_cnt) {
                max_cnt = i;
            }
            sum_cnt += i;
        }
        if (sum_cnt - max_cnt <= k) {
            return true;
        }
        else {
            return false;
        }
    }
};

File: Python3/0424_longest_repeating_character_replacement.py
# https://leetcode.cn/problems/longest-repeating-character-replacement/

class Solution:
    def characterReplacement(self, s, k):
        """
        Inputs:
            s: str
            k: int
        Outputs:
            res: int
        """

        left, right = 0, 0
        ord_A = ord("A")
        cnt = [0 for i in range(26)]
        res = 0
        while right < len(s):
            cnt[ord(s[right]) - ord_A] += 1
            while not self.isValid(cnt, k):
                cnt[ord(s[left]) - ord_A] -= 1
                left += 1

            tmp = right - left + 1
            if tmp > res:
                res = tmp

            right += 1

        return res

    def isValid(self, cnt, k):
        if sum(cnt) - max(cnt) <= k:
            return True
        else:
            return False